hooks/hookstage/invocation.go

```diff
@@ -1,11 +1,72 @@
 package hookstage
 
-import "github.com/prebid/prebid-server/hooks/hookanalytics"
+import (
+	"encoding/json"
+
+	"github.com/golang/glog"
+	"github.com/prebid/prebid-server/config"
+	"github.com/prebid/prebid-server/hooks/hookanalytics"
+	"github.com/prebid/prebid-server/metrics"
+)
+
+type Entity string
+
+const (
+	EntityHttpRequest              Entity = "http-request"
+	EntityAuctionRequest           Entity = "auction-request"
+	EntityAuctionResponse          Entity = "auction-response"
+	EntityAllProcessedBidResponses Entity = "all-processed-bid-responses"
+)
 
 type InvocationContext struct {
+	Endpoint          string
+	Stage             string
+	AccountId         string
+	DebugEnabled      bool
+	RequestTypeMetric metrics.RequestType
+	Account           *config.Account
+	moduleContexts    map[string]*ModuleContext
+}
+
+func (ctx *InvocationContext) ModuleContextFor(moduleCode string) *ModuleContext {
+	if mc, ok := ctx.moduleContexts[moduleCode]; ok {
+		return mc
+	}
+
+	emptyCtx := ModuleContext{}
+	if ctx.Account != nil {
+		cfg, err := ctx.Account.Hooks.Modules.ModuleConfig(moduleCode)
+		if err != nil {
+			glog.Warningf("Failed to get account config for %s module: %s", moduleCode, err)
+		}
+
+		emptyCtx.AccountConfig = cfg
+	}
+
+	if ctx.moduleContexts == nil {
+		ctx.moduleContexts = map[string]*ModuleContext{}
+	}
+	ctx.moduleContexts[moduleCode] = &emptyCtx
+
+	return &emptyCtx
 }
 
 type HookResult[T any] struct {
 	Reject        bool
+	NbrCode       int
+	Message       string
```

Comment:
What is `Message` and how is it different from information stored in `Errors` and `Warnings`? My original thought here is that this could potentially be a reject reason if rejecting is enabled but maybe that would already show up in `Errors`?

Reply:
We took this field from PBS-Java. The only module PBS-Java has (`ortb2-blocking`) does not provide value for this field. So we added it only for consistency, but we do not know the actual purpose of this field.
Later we add the value from this field to the `HookOutcome.Message` field, which is added as the trace to the `BidResponse` returned from the endpoint.
You can find it in the structure of the **Trace Output** returned from the PBS-Java: https://docs.google.com/document/d/1VP_pi7L5Iy3ikHMbtC2_rD5RZTVSc3OkTWKvtRS5x5Y/edit#heading=h.xbw45k8wrajs